share/sysbuild/CMakeLists.txt

```diff
@@ -94,33 +90,35 @@ add_subdirectory(bootloader)
 include(${BOARD_DIR}/sysbuild.cmake OPTIONAL)
 
 # This allows image specific sysbuild.cmake to be processed.
-list(LENGTH IMAGES images_length)
-while(NOT "${images_length}" EQUAL "${processed_length}")
-  foreach(image ${IMAGES})
-    if(NOT image IN_LIST images_sysbuild_processed)
-      ExternalProject_Get_property(${image} SOURCE_DIR)
-      include(${SOURCE_DIR}/sysbuild.cmake OPTIONAL)
-      list(APPEND images_sysbuild_processed ${image})
-    endif()
-  endforeach()
-
+set(i 0)
+while(TRUE)
+  get_property(IMAGES GLOBAL PROPERTY sysbuild_images)
   list(LENGTH IMAGES images_length)
-  list(LENGTH images_sysbuild_processed processed_length_new)
+  if(i GREATER_EQUAL ${images_length})
+    break()
+  endif()
 
-  # Check for any duplicate entries in image names to prevent an infinite loop
-  if("${processed_length_new}" EQUAL "${processed_length}")
-    # Image length was different than processed length, but no new images are processed.
-    message(FATAL_ERROR "A duplicate image name was provided, image names must be unique.")
+  list(GET IMAGES ${i} image)
+  list(FIND IMAGES ${image} j)
+  if(NOT ${i} EQUAL ${j})
+    message(FATAL_ERROR "Found multiple images named '${image}'. Image names must be unique.")
   endif()
-  set(processed_length ${processed_length_new})
+
+  ExternalProject_Get_property(${image} SOURCE_DIR)
+  include(${SOURCE_DIR}/sysbuild.cmake OPTIONAL)
+
+  math(EXPR i "${i} + 1")
 endwhile()
 
-sysbuild_module_call(PRE_CMAKE MODULES ${SYSBUILD_MODULE_NAMES} IMAGES ${IMAGES})
-foreach(image ${IMAGES})
+sysbuild_images_order(IMAGES_CONFIGURATION_ORDER CONFIGURE IMAGES ${IMAGES})
+sysbuild_images_order(IMAGES_FLASHING_ORDER FLASH IMAGES ${IMAGES})
```

Comment:
> Sure, why not?

because IMAGES_CONFIGURATION_ORDER doesn't seem to be used as a global variable, it is being (correctly) passed as argument to the `sysbuild_module_call(PRE_CMAKE ...)` function.
Secondly, the IMAGES_CONFIGURATION_ORDER  is correctly used for the `foreach()` loop, but we don't promise a specific image order being passed to modules, ref: https://docs.zephyrproject.org/latest/develop/modules.html#sysbuild-modules-hooks

Meaning users should not start to depend on a given order.
Unless there is a strong use-case for guaranteeing a specific image order for `sysbuild_module_call(...)` calls, I would prefer we don't start to give such gurantees as it might put restrictions on restructuring sysbuild code.

Also correct that IMAGES_FLASHING_ORDER  is used by domains.cmake, but why doesn't domains.cmake do the sorting itself ?

So far I see no specific argument as to why `sysbuild_module_call(PRE_DOMAINS ...)` needs the images to be ordered.

Reply:
> For simplicity, I wanted to have a single cut-off point after which it's no longer valid to add more images or update their properties.

Understandable, but that's not even the reality today.
User's can adjust any image property / cache setting until CMake configure has been invoked.
So I don't see why the image dependency should have special treatment, as it's just one setting among others.

Regarding the `PRE_DOMAINS` you may argue that we don't offer any special handling / properties dedicated to this.
Today the PRE/POST_DOMAINS hooks are provided to allow downstream users to perform custom tasks.
A use-case is merging / post processing of images before they are flashed.

However, if there is no technical reasons why users should be prevented from adjusting properties available, then I don't think we should introduce such limitation.

> because my impression was that they were mainly added to support downstream functionality like signing and such.

That is one use-case, but we don't put restrictions as to what users want to do in those hooks.

> Another concern I have is that hooks are only available for modules and not for individual sysbuild.cmake files. 

Correct, because the sysbuild.cmake serves a different purpose.
There are two kind of sysbuild.cmake files.
- Board specific
- Image specific

For the board specific one, it allows a board to adjust CMake variables, add board required images, etc.
You seem to be considering the image part, and one such image could be a board which needs to have a specific hex / bin flashed, one such example could be the Raspberry Pi Pico, ref: https://github.com/zephyrproject-rtos/zephyr/blob/3104ad0e393cc4df756b2d12aa15c0b3c9be3960/modules/hal_rpi_pico/CMakeLists.txt#L13-L15

If a board specific image is depending on user controlled images, then it's probably not so board specific, but should instead be a generally defined image which is described in Kconfig.

Similar with a sample specific image. If using `<sample>/sysbuild.cmake` to add extra images, then it should be core images to the sample. If they are general purpose samples they should be available in the modules CMakeLists txt files.
And with relevant Kconfig settings and dependencies.

So the hooks are purposely available to Zephyr modules only to help guide users into a proper CMake / Kconfig structure.
The `sysbuild.cmake` files are more specific fine-tuning / specific handling as just described.

Feel free to extend / improve our documentation, as it could be valuable to make the difference / intentions clear for the `sysbuild.cmake` files.

> Assuming that Zephyr itself is not a module, it cannot have its own hooks,

Zephyr itself is not expected to be using those hooks.
If sysbuild has special needs, then it should create a clear step, for example a built-in signing step between `CMake` and `domains`, and then one such step can provide dedicated pre/post hooks.

This is also why there is both a `PRE_DOMAINS` and `POST_CMAKE` even though POST_DOMAINS is called right after POST_CMAKE, cause that ensure users can place their processing at correct location even if more steps are added later.

>  we want that information to be more self-contained, then one option would be to support adding EXTRA_ZEPHYR_MODULES in sysbuild context, but that would require more work.

That is already supported in sysbuild, unless you refer to the NCS misuse of having samples adding themselves as a Zephyr module (which was never the intention of that setting), but guess that's another topic.

Reply:
Thanks for the detailed reply. What you wrote gave me more clarity on how this new feature should fit into sysbuild. Much of this info is indeed worth capturing in the documentation; I wouldn't mind looking into it on a separate occasion.

I've made it possible to add `CONFIGURE` deps in `PRE_CMAKE` and `FLASH` deps in `PRE_DOMAINS`.

> If using `<sample>/sysbuild.cmake` to add extra images, then it should be core images to the sample. If they are general purpose samples they should be available in the modules CMakeLists txt files. And with relevant Kconfig settings and dependencies.

That works for me, because those general purpose samples from modules are added early on (and in a well-defined order, as you mentioned elsewhere), so adding dependencies to them should be easy.

The only leftover concern I had was: what if a general purpose sample had its own core image added via `sysbuild.cmake`? Then, adding dependencies to that second image would be trickier, since it wouldn't be added at the same time as the first image. For this reason, I chose to add another patch which adjusts the image processing order.